compiler/src/iree/compiler/Codegen/Common/LinkTuningSpecsPass.cpp

```diff
@@ -61,6 +62,210 @@ static bool consumesInputOp(NamedSequenceOp op) {
+static TuningSpecsToMerge collectTuningSpecsToMerge(ModuleOp module) {
+  TuningSpecsToMerge tuningSpecs;
+  for (auto innerModule : module.getBody()->getOps<ModuleOp>()) {
+    for (auto namedSequenceOp :
+         innerModule.getBody()->getOps<NamedSequenceOp>()) {
+      if (namedSequenceOp.getSymName() != kKernelConfigSpecName ||
+          !namedSequenceOp->hasAttr(kTuningSpecEntrypointAttrName)) {
+        tuningSpecs.namedSequenceOpsToMove.push_back(namedSequenceOp);
+        continue;
+      }
+
+      auto foreachMatchOps = namedSequenceOp.getOps<ForeachMatchOp>();
+      ForeachMatchOp foreachMatch = getSingleElement(foreachMatchOps);
+      assert(foreachMatch &&
+             "ForeachMatch should exist in the `__kernel_config`.");
+
+      updateForeachMatchMappings(foreachMatch, innerModule,
+                                 tuningSpecs.namedSequenceToUser);
+    }
+  }
+
+  return tuningSpecs;
+}
+
+// Renames a `NamedSequenceOp` to resolve name conflicts caused by merging
+// tuning specs.
+static void updateNamedSequenceOp(
+    NamedSequenceOp op, OpBuilder &builder,
+    llvm::DenseMap<NamedSequenceOp, ForeachMatchOp> &namedSequenceToUser,
+    llvm::DenseMap<ModuleOp, std::string> &unnamedModuleNames,
+    unsigned &unnamedModuleCounter) {
+  StringRef specName = op.getSymName();
+  ModuleOp parentModule = op->getParentOfType<ModuleOp>();
+  assert(parentModule);
+  StringAttr parentSymbol = parentModule.getSymNameAttr();
+  std::string moduleName;
+  if (parentSymbol) {
+    moduleName = parentSymbol.getValue().str();
+  } else {
+    if (unnamedModuleNames.contains(parentModule)) {
+      moduleName = unnamedModuleNames[parentModule];
+    } else {
+      std::string newModuleName =
+          llvm::formatv("m{}", unnamedModuleCounter).str();
+      ++unnamedModuleCounter;
+      unnamedModuleNames[parentModule] = newModuleName;
+      moduleName = newModuleName;
+    }
+  }
+
+  std::string newSpecName = llvm::formatv("{}_{}", moduleName, specName).str();
+  op.setSymName(newSpecName);
+
+  // Skip updating ForeachMatchOp if the NamedSequenceOp is not used in it.
+  if (!namedSequenceToUser.contains(op))
+    return;
+
+  ForeachMatchOp foreachMatchOp = namedSequenceToUser[op];
+
+  // Helper function for updating matchers or actions in associated
+  // `foreachMatchOp` instances.
+  auto getUpdatedSymbol = [&](Attribute attr) -> SymbolRefAttr {
+    StringRef name = cast<SymbolRefAttr>(attr).getRootReference();
+    return (name == specName)
+               ? SymbolRefAttr::get(builder.getContext(), newSpecName)
+               : cast<SymbolRefAttr>(attr);
+  };
+
+  SmallVector<Attribute> updatedMatchers;
+  SmallVector<Attribute> updatedActions;
+  for (auto matcherAttr : foreachMatchOp.getMatchers()) {
+    SymbolRefAttr updatedAttr = getUpdatedSymbol(matcherAttr);
+    updatedMatchers.push_back(updatedAttr);
+  }
+
+  for (auto actionAttr : foreachMatchOp.getActions()) {
```

Comment:
Use the actual type here, since it's not obvious based on the context: https://llvm.org/docs/CodingStandards.html#use-auto-type-deduction-to-make-code-more-readable